File: internal/service/ai/mcp/model/mcp.go
package model

import (
	"time"

	"gorm.io/gorm"
)

type Tool struct {
	ID          string         `gorm:"primaryKey;size:64;comment:工具ID"`
	Name        string         `gorm:"size:128;not null;index;comment:工具名称"`
	Description string         `gorm:"type:text;comment:工具描述"`
	Type        int            `gorm:"not null;comment:工具类型 1-搜索 2-代码执行 3-天气 4-文件系统 5-自定义"`
	Config      string         `gorm:"type:text;comment:配置JSON"`
	Parameters  string         `gorm:"type:text;comment:参数定义JSON"`
	Enabled     bool           `gorm:"not null;default:true;comment:是否启用"`
	UserID      string         `gorm:"index;size:64;comment:所属用户ID"`
	CreatedAt   time.Time      `gorm:"autoCreateTime;comment:创建时间"`
	UpdatedAt   time.Time      `gorm:"autoUpdateTime;comment:更新时间"`
	DeletedAt   gorm.DeletedAt `gorm:"index;comment:删除时间"`
}

func (Tool) TableName() string {
	return "mcp_tools"
}

type ToolCallLog struct {
	ID        string    `gorm:"primaryKey;size:64;comment:调用ID"`
	ToolID    string    `gorm:"size:64;index;not null;comment:工具ID"`
	ToolName  string    `gorm:"size:128;not null;comment:工具名称"`
	Params    string    `gorm:"type:text;comment:调用参数JSON"`
	Result    string    `gorm:"type:text;comment:调用结果"`
	Status    string    `gorm:"size:32;not null;comment:调用状态"`
	CreatedAt time.Time `gorm:"autoCreateTime;comment:创建时间"`
}

func (ToolCallLog) TableName() string {
	return "mcp_tool_call_logs"
}

type MCPService struct {
	ID             string         `gorm:"primaryKey;size:64;comment:服务ID"`
	Name           string         `gorm:"size:128;not null;index;comment:服务名称"`
	Description    string         `gorm:"type:text;comment:服务描述"`
	Enabled        bool           `gorm:"not null;default:true;comment:是否启用"`
	TransportType  string         `gorm:"size:32;not null;comment:传输类型 sse/http-streamable"`
	URL            string         `gorm:"size:512;not null;comment:服务URL"`
	Headers        string         `gorm:"type:text;comment:请求头JSON"`
	AuthConfig     string         `gorm:"type:text;comment:认证配置JSON"`
	AdvancedConfig string         `gorm:"type:text;comment:高级配置JSON"`
	UserID         string         `gorm:"index;size:64;comment:所属用户ID"`
	CreatedAt      time.Time      `gorm:"autoCreateTime;comment:创建时间"`
	UpdatedAt      time.Time      `gorm:"autoUpdateTime;comment:更新时间"`
	DeletedAt      gorm.DeletedAt `gorm:"index;comment:删除时间"`
}

func (MCPService) TableName() string {
	return "mcp_services"
}

func AutoMigrate(db *gorm.DB) error {
	return db.AutoMigrate(&Tool{}, &ToolCallLog{}, &MCPService{})
}

File: web/src/api/mcp.ts
import client from './client'

function extractData<T>(res: { data: unknown }, fallback: T): T {
  const d = res.data as Record<string, unknown>
  if (!d) return fallback
  if (d.success === true && d.data !== undefined) return d.data as T
  if (d.data !== undefined && d.data !== null) return d.data as T
  if (typeof d === 'object' && !Array.isArray(d) && !('code' in d) && !('success' in d)) return d as T
  return fallback
}

function extractArray<T>(res: { data: unknown }): T[] {
  const d = res.data as Record<string, unknown>
  if (!d) return []
  if (d.success === true && Array.isArray(d.data)) return d.data as T[]
  if (Array.isArray(d.data)) return d.data as T[]
  if (Array.isArray(d)) return d as T[]
  if (d.data && typeof d.data === 'object') {
    const inner = d.data as Record<string, unknown>
    if (Array.isArray(inner.services)) return inner.services as T[]
    if (Array.isArray(inner.items)) return inner.items as T[]
    if (Array.isArray(inner.list)) return inner.list as T[]
  }
  return []
}

export interface MCPService {
  id: string
  name: string
  description: string
  enabled: boolean
  transport_type: 'sse' | 'http-streamable'
  url: string
  headers: Record<string, string>
  auth_config: Record<string, string>
  advanced_config: Record<string, string>
  created_at: string
  updated_at: string
}

export interface MCPTool {
  name: string
  description: string
  inputSchema: Record<string, unknown>
  service_id?: string
  service_name?: string
}

export async function listMCPServices(): Promise<MCPService[]> {
  try {
    const res = await client.post('/mcp/services/list')
    return extractArray<MCPService>(res)
  } catch {
    return []
  }
}

export async function getMCPService(id: string): Promise<MCPService | null> {
  try {
    const res = await client.get(`/mcp/services/${id}`)
    return extractData(res, null)
  } catch {
    return null
  }
}

export async function createMCPService(data: Partial<MCPService>): Promise<MCPService | null> {
  try {
    const res = await client.post('/mcp/services', data)
    return extractData(res, null)
  } catch {
    return null
  }
}

export async function updateMCPService(id: string, data: Partial<MCPService>): Promise<MCPService | null> {
  try {
    const res = await client.put(`/mcp/services/${id}`, data)
    return extractData(res, null)
  } catch {
    return null
  }
}

export async function deleteMCPService(id: string): Promise<void> {
  await client.delete(`/mcp/services/${id}`)
}

export async function testMCPService(data: { url: string; transport_type: string; headers?: Record<string, string>; auth_config?: Record<string, string> }): Promise<{ success: boolean; message: string }> {
  try {
    const res = await client.post('/mcp/services/test', data)
    return extractData(res, { success: false, message: '' })
  } catch {
    return { success: false, message: '连接失败' }
  }
}

export async function listMCPTools(): Promise<MCPTool[]> {
  try {
    const res = await client.post('/mcp/list')
    return extractArray<MCPTool>(res)
  } catch {
    return []
  }
}

export async function callMCPTool(data: { tool_id?: string; tool_name: string; parameters: Record<string, string> }): Promise<unknown> {
  try {
    const res = await client.post('/mcp/call', data)
    return extractData(res, null)
  } catch {
    return null
  }
}
